### ISSUE TITLE
1 minute app loading times on some Android devices

### ISSUE THREAD

**Issue by dampgnat** [airsdk/Adobe-Runtime-Support#1987]
It's likely this isn't technically a bug, but it's becoming a big problem.

Over the last 8 months my app has been live, regular content updates have accumulated APK size to 185 MB  (IPA 135 MB) and so naturally the load times have also grown.

Latest app load times:
- Samsung S21 5G - within a few seconds
- Google Pixel 4a - 30 seconds
- Redmi Note 10 Pro - 30 seconds
- Nokia G20 - 60 seconds
- Samsung Galaxy J7 Prime - 60 seconds

60 seconds is a lonnnnng time staring at a silent black screen, and unsurprisingly a lot of those users are assuming it has failed to load.

Firstly, I'm curious as to why the load time differs so much, especially when the Nokia G20 is only a year old.
All assets are included as SWCs and so must be loaded when the app loads. One developer suggested this might be audio files being unpacked. My audio SWCs are 30 MB of MP3s files. Image/animation SWCs are 55 MB.

Secondly, if nothing can be done to reduce the load time, do you know if it's possible to load assets as SWFs at runtime while retaining their class names? (iOS strips SWFs of all script including class names, so I can't even play runtime loaded sound effects).

Observed using various AIR SDKs over the last 2 years up to **33.1.1.821**

**Comment by FliplineStudios** [comment]
We've had some luck with reducing load times on Android using an old Flash trick for preloaders -- if you add a frame into the final main swf, the SWCs (and all the rest of the code and content) are moved onto the second frame, which greatly improved app startup times for us.  This was years ago, but a 60MB app with tons of content from SWCs used to take 20 seconds to start on some Samsung tablets, but by adding that frame in the beginning it dropped to only 2 or 3 seconds to start up.

We were using the Flex compiler in FlashDevelop (though it was all AS3 and SWCs), and based it on one of their preloader templates where it uses the [Frame] meta tag to add an extra frame before the rest of the content.  We were originally just doing this so we could display a "loading" screen, but it also made a huge difference with reducing the startup time!

**Comment by ajwfrost** [comment]
Hi

We've just been looking at this sort of thing for a couple of other companies, seems like a fairly common theme! so I would be very interested to know what is actually happening during this start-up time!

@FliplineStudios thanks for this info, we can have a look further to see how this idea of adding a 'preloader' into a first frame would actually happen in practice. But then, when you say it "improved app startup times", did it actually just disguise the start-up time by displaying a picture (static? or animated?), or did it really mean that your 'main' ActionScript class constructor was being called more quickly? I don't suppose you have any such SWF that's had this [Frame] injected into it so that we can see what it actually did?

We were starting to look into the idea of kicking off a separate display list within the runtime to display a preloader SWF whilst the main SWF then loaded in the background but I'm wondering whether the simpler option would be to have this just treat it like a first frame...

We can try to replicate this and see what's taking the time ... @dampgnat if there's any chance we can get your SWF file (or an APK or AAB build with `aab-debug` or `apk-debug`) we can try profiling it a little more (and can look to see whether the SWF tag ordering has got anything to do with it....)

thanks

**Comment by dampgnat** [comment]
@FliplineStudios thanks for the workaround suggestion!! Curious to know if it just offsets the loading time as Andrew suspects.
As it happens I'm also using the Flex compiler in Flash Develop, and I also used [Frame(factoryClass="Preloader")] when developing for web, but removed it for mobile.  Certainly sounds worth a try!
Did you have to change Flash/Animate SWC export settings to Export classes in frame 2?

@ajwfrost good to know I'm not the only one experiencing this. I wish I could send you a build, but my client security probably wouldn't allow it. I'll see if there's a way I can provide a less sensitive version for you to analyse.

**Comment by megajogos** [comment]
We have the same problem. It is important to say that if the user tap on the screen in this load process the ANR happens.

https://github.com/airsdk/Adobe-Runtime-Support/issues/1831

**Comment by FliplineStudios** [comment]
@dampgnat Very strange!  In our projects we still have Main.as as the main document class, and have [Frame(factoryClass = "Preloader")] in that file before the class declaration (haven't changed anything in Additional Compiler Options).  Aside from adding some text on screen, our Preloader.as is set up the same as what's generated in FlashDevelop's "AS3 Project with Preloader" template, where it checks for currentFrame == totalFrames on enterFrame, and then creates a new instance of Main after that...

**Comment by dampgnat** [comment]
Ok scrap that, we're finding that it does now seem to be loading much faster @FliplineStudios (my old Nexus 5X is shockingly inconsistent)
On a Pixel 4a - loading time has gone from an average of 25 seconds to 3! But it then gets stuck. Will continue my investigations...

**Comment by cleverbeapps** [comment]
It seems that [Frame(factoryClass = "Preloader")] does not work anymore
https://stackoverflow.com/questions/16122820/does-asc-2-0-recognize-frame-metadata-tags-ex-for-preloader-factoryclass
but instead use Additional Compiler Argument (-frame=AnyFrameLabelYouWant,YourMainClass)

**Comment by ajwfrost** [comment]
Hi

@cleverbeapps yes I found that the ASC 2.0 was ignoring it completely, but with the Flex compiler from 4.16 it worked.. thanks for the hints re. the arguments for the newer compiler!

This is quite an interesting one, looking at what's going on under the hood. The inclusion of the preloader will mean that as soon as the first 'frame' is reached, the content will start displaying, and this content is just using the preloader class as if the whole application was the preloader. I'm actually a bit surprised that this works as I thought it was all single-threaded, but it looks like the file is loaded in chunks via a background thread. So the runtime will continue to process the SWF as it comes in, until it's fully loaded at which point the preloader is complete.

The challenge I then see though is that the preloader (or at least, the template here from FlashDevelop) is doing an 'addChild' with a new symbol which is the actual main document class. Which means that (a) the main document class can't access the stage within it's constructor, and (b) there is an extra set of objects in the hierarchy which can lead to scaling/resizing issues when things like orientation changes happen.

I would actually have expected there to be the initial frame containing the preloader, then a second frame containing the main application class, so the preloader doesn't have to add the main app as a child. But, the preloader is set as the "main" application class (SymbolClass tag, ID of 0) which means it's associated with the overall root timeline. Will have to see if we can craft a SWF that has a preloader on frame one added via PlaceObject, and then a main app class added just in the second frame... 

@dampgnat just in terms of your debugging here, are you using Scout? as that may help particularly if there are exceptions being thrown internally that are causing it to stop..

thanks

**Comment by cleverbeapps** [comment]
@ajwfrost  Hi Andrew. 
I have tried (IntellijIdea IDE):
1) in Additional Compiler Options added -frame=ContentFrame,Main
2) changed my main application class link to my custom Preloader.as (from https://gist.github.com/aajanki/751852)
3) in that Preloader.as on line 33 (nextFrame()) - application is stopped and have no response. I have tried goToAndStop("ContentFrame"), but nothing changed.

Could you share how to get it work? Thanks.

**Comment by dampgnat** [comment]
@cleverbeapps for some reason I had to move my Main class into my domain and include that to make it work.

Additional Compiler Options: -frame start com.company.project.Main

In my Preloader.as I wait until all bytes are loaded and then:
```
var mainClass:Class = getDefinitionByName("com.company.project.Main") as Class;
addChild(new mainClass() as DisplayObject);
```

@ajwfrost I have used Scout quite a bit in the past but only for isolating performance bottlenecks. My client's SDK makes it difficult to do debug builds. The device that did run the APK without freezing had focus problems when reopening it (not touch response), perhaps that's related to the stage access issue you mentioned?
I'm going to see if I can remove some stuff and then send you an APK as you requested. Do you have an upload link?

**Comment by ajwfrost** [comment]
@cleverbeapps  in your preloader code, you say it's hanging on `nextFrame()`  - but I'm not sure why you'd want to change frames here at all... in the FlashDevelop template it calls `stop()` there...

@dampgnat interesting, I think there are some issues with the display list ending up unnecessarily complicated. I just tried changing the preloader to:
```
		private function startup():void 
		{
			var s : Stage = stage;
			var mainClass:Class = getDefinitionByName("Main") as Class;
			s.addChild(new mainClass() as DisplayObject);
			s.removeChild(this);
		}
```
which I think simplifies it a bit...

For uploads: https://transfer.harman.com/requests/dp5JuDJjFPJ0donrpcPemS
thanks

**Comment by dampgnat** [comment]
@FliplineStudios heheh... that ol' chestnut :)
Yeah, that's in my code, and it's working fine on some Android devices, but not others. Still investigating...

**Comment by FliplineStudios** [comment]
@dampgnat Interesting, I know we've stuck with using swf-version 37 in our own projects, because 38 was when Adobe switched PC builds to use Direct3D 11 instead of Direct3D 9, and for some reason AIR thinks all of my PCs don't support DirectX 11 so it falls back to software rendering.  That's crazy that it gets even worse with the Preloader setup!

**Comment by dampgnat** [comment]
@FliplineStudios hey, are you suggesting that AIR now harnesses the GPU for PC builds?? Or is that for stage3D/Starling only?

**Comment by FliplineStudios** [comment]
@dampgnat I'm not entirely sure, at least it doesn't seem exactly the same as GPU mode on mobile -- but I do notice a performance increase on PC when setting renderMode to gpu/direct even with classic display list vector graphics, compared to cpu renderMode being sluggish.

**Comment by megajogos** [comment]
Please anyone knows how to do that inside Animate? How to specify this custom preloader?

**Comment by cleverbeapps** [comment]
@megajogos honestly, using if Adobe Animate for AIR development is kind of rudimental, like Flash Builder. Maybe I am wrong, but it is pain to work with that IDE etc. From me, I can suggest switching to InteligIDEA, and life will be much easier.

**Comment by megajogos** [comment]
We don't use Animate as IDE, but our CI uses Animate to build the swf and ADT to build the apps and we have almost 40 apps, so this CI is very important.

@ajwfrost do you know how to specify this custom preloader inside the Animate?

**Comment by ajwfrost** [comment]
@megajogos if you're using Animate anyway, which will generate a SWF with a main application already based on MovieClip, then you can set this up manually I would think? So just add a keyframe to the very start of your main timeline and set the document class to be the preloader .as file...

**Comment by ajwfrost** [comment]
@dampgnat belated response (I wrote it yesterday but forgot to hit the 'comment' button!)

I was going to ask, what renderMode are you running it in? cpu would always use cpu, but auto or gpu (or direct) would try to set up Direct3D, and in general it tries with Direct3D 11 now .. but yes the fallback appears to be to cpu rendering rather than to Direct3D 9.

Not sure why the use of a preloader makes any difference... we can look into that.

**Comment by cleverbeapps** [comment]
@ajwfrost could you please give a link to ASC2.0 documentation that describes all compiler arguments, if there any?

**Comment by dampgnat** [comment]
@ajwfrost interesting, my desktop project appears to have been running with GPU anyway, as my application descriptor didn't specify a renderMode and it clearly runs slower if I specify "cpu".

As a side note, I had no idea the application descriptor file was used when building/testing the SWF (I use Flashdevelop). I assumed that was only for packaging (like when packaging for mobile).

I've discovered that Flashdevelop is what's causing the massive performance hit when running in Debug mode (much worse with preloader), as performance returns to normal when I close Flashdevelop while testing the SWF. I suppose this kind of thing is to be expected when using unsupported software.


Regarding the APK loading time issue - have there been any interesting findings with the APKs I sent?

**Comment by dampgnat** [comment]
Hi @ajwfrost were the APKs I sent of any use? Or do you also need the SWF for analysis?
The loading time issue has become a blocker for our client, so we're very keen to find a way a forward.

**Comment by ajwfrost** [comment]
Hi @dampgnat  - apologies for the delay, we've had some other issues taking up a bit too much time to check on this. We're actually looking into building a preloader type capability into the runtime which should then help with masking the start-up but also depending on how it works, it may improve the overall time per the comments earlier here..

We can pull the SWF out of the APK though :-) so we have everything we need I think now.

thanks!

**Comment by dampgnat** [comment]
Thanks for the update @ajwfrost! Fingers crossed that avoids the side effects caused by the current preloader workaround.
Let me know if you need beta SDKs tested this end.

**Comment by dampgnat** [comment]
Hi @ajwfrost I noticed 33.1.1.926 was released last week but with no mention of this issue. Is there a visible road-map for planned fixes? We're trying to manage client expectations on this.

**Comment by ajwfrost** [comment]
Hi @dampgnat  - yes, sorry nothing in the 33.1 builds on this, we’re looking at it under our “future” branch for which we’ll get a beta out soon (I’ve been a saying that for a while… but it’s basically ready, just pushed back by the need for another 33.1 update)

but in terms of when the preloader improvements will be in, we’re probably looking at the end of September now.

thanks

**Comment by ajwfrost** [comment]
Hi all - just to provide a bit of an update here. We've been investigating a few different options for how to properly create preloaders, and trying to ensure we solve any issues with odd handling of device orientation, scaling etc when you load in the main class. 

Mostly the problems are caused here because in all these solutions the 'main' Sprite-based application is being added as a child (either of the stage or of the preloader) but it's being created first, and there are some properties that are set up when it's created which are then seemingly immutable. These then cause problems if you actually have dynamic layout code within this, so e.g. there's a fixed relationship between the Sprite's "height" and "scaleY" properties. I think a lot of the normal application layout code depends upon the fact that the main Sprite-based class that is created at start-up is handled in a slightly special way as it's designated as the "root" element.

So the way we've managed to get this working is to force a preloader to be completely removed from the display list and for that to be reset, with the main application then being created whilst it's already present on the stage as the root element, just as if it was created normally.

This I think provides the best approach/results. It should mean that you can develop your application like normal and do whatever code you think best for it, and then bolt on a preloader separately a bit later on. The preloader can listen out for the `loaderInfo` progress and when it's reached 100% it would need to call a new method to trigger the switch over to the main class, currently we've just created `stage.preloadComplete("Main");` where you give it the name of the class that it should then set up as the main application. (We wondered about just automating that whole process - i.e. you wouldn't have to do anything, as soon as the file was fully loaded we would automatically find and switch to that class - but I'm not sure that would be such a good idea... and we may be able to have a way of working out the 'main' class name so that you could reuse a preloader between apps, but I suspect that wouldn't be something people would do anyway..)

So far, those changes in the runtime are fairly small/self-contained: but the challenge then is, to make this work we need to merge in the preloader to the start of the main SWF. We'd tried keeping them separate but it just raises loads of new issues/problems.. For testing, we are using a Flex compiler with the ActionScript main class having a directive:
```
	  [Frame(factoryClass = 'Preloader')] //class name of your preloader
```
at the start of it. I believe the same can be achieved with a command-line option for the compiler. But, those are currently just ignored by the ASC 2.0 compiler that's used in normal AIR projects...

We then have two options:
1) try to update the ASC 2.0 compiler so that it builds in the 'frame' capability again to match what the Flex SDK compiler does
or
2) create a separate utility that would merge things at the SWF level i.e. you build a preloader SWF, and you build your main application, and then you run a utility that turns these two into a single SWF that contains both of them.
or both!

Option 2 is relatively straightforward I believe; option 1 we would have to look into but should be possible I hope.. but may be something we do as a second stage.

Would welcome any feedback on the concepts/approach here :-) and will try to get this into our forthcoming beta SDK.

thanks

**Comment by dampgnat** [comment]
Thanks for the update on this @ajwfrost!  Your bolt on preloader approach sounds ideal/safe.

I'm all for automation if it works as expected, but also assume that some devs would like some control over that switch over. Maybe it depends on whether a visual loading screen is dependent on the preloader phase. Some might prefer user control over that switch? But then I guess there are ways of cheating that, so not vital.

I'm curious about what "...run a utility that turns these two into a single SWF..." entails.
I'm not thrilled by the prospect of on an additional step to the packaging process, but at this stage - whatever it takes for an un-bottlenecked loading time. Would be interesting to hear others weigh in on this.

**Comment by ajwfrost** [comment]
Re the merging of two SWFs .. that would just be a standalone tool (AIR-based, naturally) that merged two single-frame SWF files into a SWF with two frames. In theory, that's relatively simple.

I'm actually thinking that doing this in the compiler may not be too tricky and I think that would be better (especially if we can use that `[Frame]` tag inline in the AS3 file). Since I also agree, adding a separate step would be a pain..

thanks

**Comment by dampgnat** [comment]
Hi @ajwfrost congrats on the AIR 50 BETA launch!  I've already tested an Android build using this and all appears well.

I've been trying to get the new preloadComplete working today. As a few approaches were discussed above, I'm not entirely sure I'm doing it right. This is what I've attempted with FlashDevelop using SDK 50.0.0.1 and -swf-version=50

Preloader.as is set as the Document Class
When the preloader class has detected that all bytes are loaded I call `stage.preloadComplete("Main")` :thumbsup:
"Main" wasn't being found until I added additional compiler options `-frame start Main` :thumbsup:
Tracing `stage.preloadComplete("Main")` returns `true` and the preloader class appears to have been removed. :thumbsup:
but the Main class doesn't appear to run and no further traces are called. :confused:

**Comment by ajwfrost** [comment]
@dampgnat sorry for the delay .. will have to check on this one. Are you using a Flex SDK that's overlaid with the AIR SDK, as I think this is what would be needed for now...?

**Comment by dampgnat** [comment]
@ajwfrost correct - I am indeed using the Flex SDK overlaid with the AIR SDK.  In general I've been overlaying each new AIR SDK on top of the same FlexSDK for years (version 4.6.0).

**Comment by megajogos** [comment]
@ajwfrost we would like to use the "delayed-load capability for ANEs". Could you please describe how to use it?

**Comment by marchbold** [comment]
@megajogos We are doing some research into this atm and hope to have some info / docs about it soon. We are aiming to have simple wrappers for all our extensions to use this functionality, hopefully making the integration straight forward.

**Comment by ajwfrost** [comment]
From the release notes:
> AIR-5960: Adding delayed-load capability for ANEs
> To avoid having to load and initialise everything within the first frame of execution, ANEs can now be
loaded in a delayed fashion. To set this up, add a `delayLoad="true"` attribute into the
“extensionID” tag, and then – prior to accessing any of the AS3 definitions from the ANE’s library –
use the new `ExtensionContext.loadExtension()` method. This method can also be used to load an
ANE that’s not listed in the application descriptor file.

That last sentence isn't fully true with the pre-release version of 50.0 but is completed now so would be available in our next release.

**Comment by marchbold** [comment]
@ajwfrost Did that "is loaded" function get added?

**Comment by ajwfrost** [comment]
Ah, I'd forgotten about exposing the "is loaded" state... 

Currently it's all synchronous - basically the same as the process that normally happens at app start-up - which means that it may be worth loading one per frame. The actual loading process consists of (a) finding/loading the bytes from the library.swf, and (b) parsing this library into the AS3 engine so that it's available when applications use those symbols. Then loading in the native library (for C-based ones) happens when you first create the extension context (which is done as a separate step from the loadExtension() call... but if you try creating an extension context for a not-yet-loaded extension, we'll try to load it first...)

If you call "loadExtension" it will return true if the extension is loaded.. if it's already loaded and you call it again, it will still return true.

Hope that helps!

**Comment by marchbold** [comment]
Ah no problems, we can add something to the helper I'm creating to handle a loaded state.

**Comment by marchbold** [comment]
One other question is there a way to check if an extension has the "delayLoad" flag set?

**Comment by ajwfrost** [comment]
That's another good idea....! (not without starting to parse the application descriptor file in the AS3 code, which isn't the easiest of things to do...)

Just thinking about these; we don't at this point have an ExtensionContext instance, so if you're trying to get properties about an extension, you'd perhaps need to have a way to query this with the extension IDs. Or maybe more generically...? would it be useful to have a static property on the `ExtensionContext` class such as `extensionDetails` which listed all of the ANEs (i.e. any found in the app descriptor file, plus any that had been dynamically loaded since then), each of which then has properties such as `isLoaded`, `isDynamic`, `inAppDescriptor` - meaning if it's loaded or not; if it's loaded on demand/dynamically (so 'delayLoad' causes this to be true); and if it's in the app descriptor vs if it's been loaded without any such reference (which is the new capability just being added now).

So e.g. new class 'ANEInfo' with these properties alongside the extension ID of course (and a `load()` method?). We could also add other properties such as the available function names, plus the version from the extension descriptor xml?

Let me know what would be useful :-)  these sorts of things are relatively straightforward to implement now..

**Comment by megajogos** [comment]
@marchbold @ajwfrost Any news about this? We really would like to use this dalayed-loaded capability.

**Comment by ajwfrost** [comment]
@megajogos so we're just finalising the updates for this, which would then allow Distriqt et al to add the necessary checks and capabilities to the ANEs and documentation. Current plan is for the next release to come out around the middle of next week ..

**Comment by dampgnat** [comment]
Hi @ajwfrost, good to see today's official release of v50.

We've pounced on this, but still can't get Main.as to run after using the new preloader functionality and can't find any new information on this in the release notes or dev portal.
Because of the 1 minute loading times - our client have been waiting for the app update since June, so we're _really_ under a lot pressure.

We'd really appreciate some guidance on this, we've not been able to progress from here - https://github.com/airsdk/Adobe-Runtime-Support/issues/1987#issuecomment-1249556574

**Comment by ajwfrost** [comment]
@dampgnat let me double-check this .. from your earlier description it looks like it should all be set up properly, I'm wondering if there's a reference error happening though. Are you running it on an Android device, or on Windows/Mac within the debug launcher? and are you seeing anything in Scout, sometimes that would show up a verify error in the log..

I'm wondering whether it's possible for us to check your SWF file to see what's actually happening internally; if that's possible please let me know (or just send it to me if you can) or if it's possible to create a minimal test case without any private stuff in it, that may also help.

thanks

**Comment by dampgnat** [comment]
Hey @ajwfrost, I'm running on Windows within the debug launcher.
No logs seem to appear in Scout, perhaps I need advanced telemetry? Not sure I can add that in Flashdevelop.

I've put the preloader into an old project as a minimal test case and output a SWF ready for you. I was curious what would happen if I opened that SWF directly and to my surprise the flashplayer reports the reference error you anticipated. Strange that the debugger didn't receive these in the trace output when building from Flashdevelop.  Anyway here's the report:
`ReferenceError: Error #1069: Property preloadComplete not found on flash.display.Stage and there is no default value.`

Can you send a transfer request so I can send you the SWF?

**Comment by ajwfrost** [comment]
Hi @dampgnat  - something a bit odd if nothing appears in Scout at all. Advanced telemetry shouldn't change what I was looking for there (but see the project properties, "Compiler options" tab on the right, you can change the advanced telemetry setting there).

The reference error when running in the Flash Player projector/debugger is just because we're using a new API that was added by Harman, the Flash Player doesn't have that API (`Stage.preloadComplete()`) so the reference error happens there. (ah, yes, just spotted your edit in the above -> yes exactly). But within that call, when it succeeds i.e. when running in AIR 50.0, it sounds like you were getting into that and it was returning true. I think actually if we were not finding the `Main` class, we would probably have thrown a reference error at that point too... so if it's returning a value then it implies it's successfully created the new object...

We can check though, so please see below for an upload link for the SWF:
https://transfer.harman.com/requests/rCiPYIy0sjjUT0KTriPHhz

thanks

**Comment by dampgnat** [comment]
File sent!

Sorry, I just meant no "trace" logs in Scout. Plenty of other data.
And yeah if I misspell "Main" when calling preloadComplete it does indeed throw a reference error.

**Comment by megajogos** [comment]
@marchbold @ajwfrost Any news about this topic?

**Comment by dampgnat** [comment]
Hi @ajwfrost our client's QA have confirmed that the total Android loading times (using the new preloaderComplete in SDK 50.0.1.3 discussed in https://github.com/airsdk/Adobe-Runtime-Support/issues/2207 are indeed significantly reduced. On the devices where it was taking 30 seconds, it now only takes 2 to 3 seconds!!

This is not just a HUGE improvement in UX, but surely an ESSENTIAL one. I implore that you highly recommend this approach to all Android AIR developers. Thank you so much for integrating this solution Andrew, our client is very happy!

**Comment by ajwfrost** [comment]
Thanks @dampgnat that's good to hear! We'll continue to look into improving the preloader and start-up capabilities, per the current conversation under #2207 

cheers

**Comment by Ender22** [comment]
> Hi @ajwfrost our client's QA have confirmed that the total Android loading times (using the new preloaderComplete in SDK 50.0.1.3 discussed in #2207 are indeed significantly reduced. On the devices where it was taking 30 seconds, it now only takes 2 to 3 seconds!!
> 
> This is not just a HUGE improvement in UX, but surely an ESSENTIAL one. I implore that you highly recommend this approach to all Android AIR developers. Thank you so much for integrating this solution Andrew, our client is very happy!

Hey sorry, can I ask, in what kind of app setups will this have an effect? Does it apply to all apps or just app made in a certain way like non-starling apps or something? 
I don't really understand why your apps would have 30 second loading times before this change, are you loading all assets into memory at once immediately rather than loading sounds or sprite sheets or ANEs etc during a splash screen or something? Just wondering if this change applies to my app or not.

**Comment by dampgnat** [comment]
Hi @Ender22 I reckon this affects _all_ apps, on _some_ Android devices (and not just old ones - eg. Nokia G20 was only last year), and to the extent relative to SWF file size.
The QA teams I used obviously can't test all, but found it was affecting 6 of their devices, so potentially a _lot_ more.

In my case - the SWF size is about 90MB (on a Nokia G20 it was taking 1 minute to load!), the bulk of which is sound files and bitmaps in SWC files included within the single SWF generated by AIR which loads when the app is started. Nothing extra is loaded on the fly.

Early in this thread Andrew observes of AIR - "...it looks like the file is loaded in chunks via a background thread", so it could be that some Android devices can't handle the multi-threading, causing the load process to congest/bottleneck.

FliplineStudios also recollects how they inadvertently discovered a significant load time reduction when implementing the _old_ preloader. So the issue has existed for quite some time.

I would suggest finding access to a problematic device (I've listed 4 at the top of this thread, maybe try any Pixel 4) and then running some tests. And although it may not be anywhere near as bad on other Android devices - the preloader could still have a significant impact on loading time.
